### PR TITLE
Improve generation of unique strings

### PR DESCRIPTION
**Pull request by davidbolvansky**

### CODE REVIEW

File: lib/cstring.h [hunk 1]
```diff
@@ -291,12 +291,13 @@ inline std::string& operator+=(std::string& a, cstring b) {
     return a; }
 
 template<class T> cstring cstring::make_unique(const T &inuse, cstring base, char sep) {
-    char suffix[12];
     cstring rv = base;
-    int counter = 0;
+    static boost::uuids::random_generator generator;
```
[the remark this answers is not in the document]
Comment (davidbolvansky) [reply lib/cstring.h:295]: Any suggestion to avoid this O(n) loop?

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply lib/cstring.h:295]: Do you have evidence that it's a problem in practice?

[the remark this answers is not in the document]
Comment (davidbolvansky) [reply lib/cstring.h:295]: Yeah, we see it in perf profile..

[the remark this answers is not in the document]
Comment (davidbolvansky) [reply lib/cstring.h:295]: static int counter possibly..

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply lib/cstring.h:295]: A simple solution would be to remember the last generated id and start there. Just make the counter 'static'.

**Comment by mbudiu-vmw** [comment]
This change will cause hundreds of files generated by the compiler to change, cause the corresponding tests to change.
You need to submit these files as well.

**Comment by davidbolvansky** [comment]
> This change will cause hundreds of files generated by the compiler to change, cause the corresponding tests to change. You need to submit these files as well.

Is there any way how to autoregenerate outputs for failed tests?

**Comment by vhavel** [comment]
@davidbolvansky re-run failed tests (or whole test-suite) with `P4TEST_REPLACE=True`

We have just seen real P4 program which takes hours to compile. The source program is large, and is additionally inflated by some midend transformations used in our backend. The typeChecking on huge IRs then becomes realy slow.

If we don't want to change all the identifiers, we could also implement the unique string generation in typeChecking manually. Maintaining `inUse` set and `counter` in different functions seems to prevent the unique string generation to be implemented efficiently and cleanly.

**Comment by davidbolvansky** [comment]
Ok, to reduce test churn, I implemented other approach based on usage of references to counters, as suggested by Vojta.

For every name, we track how many times it was used as a "base" for newly generated strings.

**Comment by mbudiu-vmw** [comment]
Indeed, the compiler has not really been designed to be fast. I am not sure how much we can fix this now.
If this is ready for review please ask for one.
You need to document the new map structure, though.

**Comment by davidbolvansky** [comment]
Yeah, I will document it soon.

**Comment by davidbolvansky** [comment]
> Indeed, the compiler has not really been designed to be fast. I am not sure how much we can fix this now.

Understood, but there are low hanging fruits (like this one) which can improve compile times a lot. We have a program where better unique string generation improves compile time significatly (2-3 hours -> 5 min).

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**